library/net.c

```diff
@@ -1208,12 +1208,14 @@ struct dnet_net_state *dnet_state_create(struct dnet_node *n,
 err_out_send_destroy:
 	pthread_mutex_lock(&n->state_lock);
 err_out_unlock:
-	dnet_state_rb_remove_nolock(st);
-	list_del_init(&st->storage_state_entry);
+	dnet_state_remove_nolock(st);
 	pthread_mutex_unlock(&n->state_lock);
+
 	dnet_state_put(st);
-	pthread_mutex_destroy(&st->send_lock);
-	pthread_mutex_destroy(&st->trans_lock);
+	dnet_state_put(st);
```

Comment:
It's undetermined what refcnt will left after `dnet_state_move_to_dht` call. In case of error, `dnet_state_move_to_dht` can put (`-EEXIST`) or not to put (other errorcode) the net state. So you can't know how many puts you must do at this line.

I think the good practic is to do like `shared_ptr`: we lock the resource in the component where we use this resource (in `dnet_state_create`), and don't do refcounting for another components. These another components must do clear refcounting by themself. For example: `dnet_state_create` must have scoped single refcnt-lock, and `dnet_state_move_to_dht` should inc (if success) or don't inc (if fail) refcnt. The same way in `dnet_state_create`: if success, we return net state and don't decrement refcnt; if fail, decrement refcnt and return NULL. The caller of `dnet_state_create` must keep the result object, or throw it away via put.